File: 1387-sort-integers-by-the-power-value/1387-sort-integers-by-the-power-value.py
class Solution:
    def getKth(self, lo: int, hi: int, k: int) -> int:
        @cache
        def solve(x):
            if x == 1:
                return 0
            elif x % 2 == 0:
                return 1 + solve(x // 2)
            else:
                return 1 + solve(3 * x + 1)
        arr = []
        for i in range(lo, hi + 1):
            arr.append([i, solve(i)])
        arr.sort(key=lambda x: x[1])
        return arr[k - 1][0]

File: 1387-sort-integers-by-the-power-value/1387-sort-integers-by-the-power-value.cpp
class Solution {
public:
    unordered_map<int, int> m;
    int solve(int x){
        if(m.count(x)){
            return m[x];
        }
        else if(x % 2 == 0){
            return m[x] = 1 + solve(x / 2);
        }
        else{
            return m[x] = 1 + solve(3 * x + 1);
        }
    }
    int getKth(int lo, int hi, int k) {
        m[1] = 0;
        vector<pair<int, int>> v;
        for(int i = lo; i <= hi; i++){
            v.emplace_back(make_pair(i, solve(i)));
        }
        sort(v.begin(), v.end(), [&](pair<int, int > x, pair<int, int> y){
            return x.second < y.second || (x.second == y.second && x.first < y.first);
        });
        return v[k - 1].first;
    }
};